packages/firestore/lib/index.d.ts

```diff
@@ -1969,6 +1969,24 @@ export namespace FirebaseFirestoreTypes {
      * ```
      */
     clearPersistence(): Promise<void>;
+    /**
+     * Waits until all currently pending writes for the active user have been acknowledged by the
+     * backend.
+     *
+     * The returned Promise resolves immediately if there are no outstanding writes. Otherwise, the
+     * Promise waits for all previously issued writes (including those written in a previous app
+     * session), but it does not wait for writes that were added after the method is called. If you
+     * want to wait for additional writes, call `waitForPendingWrites()` again.
+     *
+     * Any outstanding `waitForPendingWrites()` Promises are rejected during user changes.
```

Comment:
This part seems to show how we handle their documentation line here: "Any outstanding waitForPendingWrites() Tasks are cancelled during user changes."

So if they cancel the task, it bubbles up to us as task completed with a cancel, and we reject? Is there a way to differentiate between rejected-because-something-terrible-happened and rejected-because-a-new-write-came-in? Those two cases seem important to handle differently and will likely be the first thing users ask for support on. Documenting it specifically enough for users to handle based on already-available information would save maintainer+dev time with an issue back and forth I think. Bonus points for adding an e2e test that exercises this API and attempts to actually differentiate them

What do you think?

Reply:
It actually rejects if the user logs out and back in, not because of a write (this was unclear to me too, and I've just made the documentation clearer).

I've added this scenario into the e2e test, though I'm not sure how to add the rejected-because-something-terrible-happened scenario.

Reply:
Fascinating - that is not what I was expecting based on my first read of the docs - the added test that probes it and the docs are great now, given that testing result. Can't tell you how much seeing a PR with tests (and then an iteration, that improves the docs) is just amazing, lends confidence to releasing it. Thanks

I'll double check the version stuff (just to make sure we don't accidentally issue a breaking change without a semver major - no problems either way I just want to get that right, no problem issuing a semver major...)

This looks great though

Reply:
Oh and specifically, I'm not sure how to test the something-terrible-happened either, that usually is something we learn after it's in the wild, and add cases as needed